src/cppsim/circuit_optimizer.hpp

```diff
@@ -34,16 +34,31 @@ class DllExport QuantumCircuitOptimizer {
     ////////////////////////////////////////////////////////////
     // for swap insertion
     ////////////////////////////////////////////////////////////
+    struct GateReplacer {
+        std::map<QuantumGateBase*, QuantumGateBase*> replace;
+
+        QuantumGateBase* get_replaced_gate(QuantumGateBase* gate) {
+            if (!replace.count(gate)) return gate;
+            return replace[gate] = get_replaced_gate(replace[gate]);
+        }
+
+        void set_replaced_gate(QuantumGateBase* from, QuantumGateBase* to) {
+            replace[from] = to;
+        }
+    };
```

Comment:
Optimization by swap insertion modify target qubit indices but manage gates by pointer.
New function `create_gate_whose_qubit_indices_are_replaced` changes gate pointer, so I created this `replacer`, manage replacement by referencing it, and actually replace gates later.